File: 39-combination-sum/39-combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
  n := len(candidates)
  ans := make([][]int, 0)
  sort.Ints(candidates)
  var backtrack func(path []int, i, target int)
  backtrack = func(path []int, i, target int) {
    l := len(path)
    for ; i < n; i++ {
      v := candidates[i]
      if target < v { return }
      if target == v {
        res := make([]int, l+1)
        copy(res, path)
        res[l] = v
        ans = append(ans, res)
        return
      }
      path = append(path, v)
      backtrack(path, i, target - v)
      path = path[:l]
    }
  }
  backtrack([]int{}, 0, target)
  return ans
}

File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
  const ans = []
  const backtrack = (cand, stack, sum) => {
    for (let i = 0; i < cand.length; i++) {
      const v = cand[i]
      const next = sum + v
      if (next < target) {
        backtrack(cand.slice(i), [...stack, v], next)
      } else if (next === target) {
        stack.push(v)
        ans.push(stack)
      } else return
    }
  }
  candidates.sort((a, b) => a - b)
  backtrack(candidates, [], 0)
  return ans
};